pingcap/tiflow
# checker(dm): change checker to more smart and gentle

### What problem does this PR solve? 

Issue Number: close #3766

### What is changed and how it works?

1. https://github.com/pingcap/tiflow/blob/1fb41f91bb048860c924b4752dee4f22137635d2/dm/docs/RFCS/20211130_enhanced_pre_checker.md?plain=1#L61-L75

2. add binlog_do_db checker

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

#### test case-sensitive

1. start MySQL with 
```cnf
[mysqld]
...
binlog_do_db = Upper_DB
binlog_do_db = Upper_DB1
lower_case_table_names = 1
```
```sql
mysql> show master status;
+------------------+----------+--------------------+------------------+-------------------------------------------+
| File             | Position | Binlog_Do_DB       | Binlog_Ignore_DB | Executed_Gtid_Set                         |
+------------------+----------+--------------------+------------------+-------------------------------------------+
| mysql-bin.000006 |     3249 | Upper_DB,Upper_DB1 |                  | a0ec2f1a-d811-11eb-84a7-16210329b3ba:1-52 |
+------------------+----------+--------------------+------------------+-------------------------------------------+
```
2. create table in upstream
```sql
create database `Upper_DB`;
create database `Upper_DB1`;
mysql> show databases;
+-------------------------+
| Database                |
+-------------------------+
| upper_db                |
| upper_db1               |
+-------------------------+
```
3. start dm-master/dm-worker and create source
4. check task with 
```yaml
case-sensitive: false

mysql-instances:
  - source-id: "mysql-replica-01"
    black-white-list:  "instance" # compatible with deprecated config
    mydumper-config-name: "global"
    loader-config-name: "global"
    syncer-config-name: "global"
    route-rules: ["global"]
    filter-rules: ["global"]

black-white-list: # compatible with deprecated config
  instance:
    do-dbs: ["Upper_DB*", "lower_db"]
    do-tables:
      - db-name: "Upper_DB*"
        tbl-name: "Do_Table*"
      - db-name: "Upper_DB"
        tbl-name: "lower_table*"
      - db-name: "lower_db"
        tbl-name: "Upper_Table*"

routes:
  global:
    schema-pattern: "Upper_DB"
    table-pattern: "Do_Table"
    target-schema: "UPPER_DB_ROUTE"
    target-table: "do_table_route"
```
5. check pass

#### test check failed(not in binlog_do_db)

1. If remove `Upper_DB1` from `binlog_do_db`
```sql
mysql> show master status;
+------------------+----------+--------------+------------------+-------------------------------------------+
| File             | Position | Binlog_Do_DB | Binlog_Ignore_DB | Executed_Gtid_Set                         |
+------------------+----------+--------------+------------------+-------------------------------------------+
| mysql-bin.000007 |      194 | Upper_DB     |                  | a0ec2f1a-d811-11eb-84a7-16210329b3ba:1-52 |
+------------------+----------+--------------+------------------+-------------------------------------------+
1 row in set (0.00 sec)
```
2. check failed.
```
» check-task /home/xianjie/ticdc/dm/tests/case_sensitive/conf/dm-task.yaml
{
    "result": false,
    "msg": "[code=26005:class=dm-master:scope=internal:level=medium], Message: fail to check synchronization configuration with type: check was failed, please see detail
        detail: {
                "results": [
                        {
                                "id": 7,
                                "name": "binlog_do_db/binlog_ignore_db check",
                                "desc": "check whether dbs need replication is in binlog_do_db/binlog_ignore_db",
                                "state": "fail",
                                "extra": "these dbs [upper_db1] are not in binlog_do_db[upper_db]"
                        }
                ],
                "summary": {
                        "passed": false,
                        "total": 9,
                        "successful": 8,
                        "failed": 1,
                        "warning": 0
                }
        }"
}
```

Related changes

 - Need to update the documentation

### Release note 

```release-note
Please add a release note.
If you don't think this PR needs a release note then fill it with `None`.
```
